### PR TITLE
ISPN-12881 Rolling Upgrade configuration can not removed from a dynamic created cache after migration

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.redhat.com/browse/ISPN-12881

Add support for doing Rolling Upgrades without needing to change cache's configuration to add a remote store statically, adding and removing it at runtime. 

~~Marking as preview as it only contains the core/persistence changes~~

### CODE REVIEW

File: core/src/main/java/org/infinispan/factories/EmptyConstructorNamedCacheFactory.java [hunk 1]
```diff
@@ -122,14 +121,11 @@ public Object construct(String componentName) {
       } else if (componentName.equals(CommandsFactory.class.getName())) {
          return new CommandsFactoryImpl();
       } else if (componentName.equals(PersistenceManager.class.getName())) {
-         if (configuration.persistence().usingStores()) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/factories/EmptyConstructorNamedCacheFactory.java:125]: There are more references to `usingStores` that need to be updated. `ActivationManagerImpl`, `PassivationManagerImpl` are just a couple as they check if stores are configured at startup.

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java [hunk 2]
```diff
@@ -596,6 +623,14 @@ private long getFlagsForStateInsertion() {
       return aggregateCompletionStage.freeze();
    }
 
+   @Override
+   public void addStore(PersistenceConfiguration persistenceConfiguration) {
+      createAndStartStores(persistenceConfiguration);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:628]: Another problem is that a user may think they can change if passivation is enabled or not since they can pass a `PersistenceConfiguration`. I think this would be better if instead the argument is just a `StoreConfiguration`

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java [hunk 3]
```diff
@@ -596,6 +623,14 @@ private long getFlagsForStateInsertion() {
       return aggregateCompletionStage.freeze();
    }
 
+   @Override
+   public void addStore(PersistenceConfiguration persistenceConfiguration) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:627]: The `disableStore` method needs to expand its locking now that a store can be added at runtime. We need to only update the interceptor chain in the writeLock to prevent concurrent updates to it.

File: documentation/src/main/asciidoc/topics/config_examples/remote-store.json [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{
+  "remote-store": {
```
Comment (wburns) [inline documentation/src/main/asciidoc/topics/config_examples/remote-store.json:2]: Serializing the entire persistence configuration seems overkill. Can we not just serialize a RemoteStoreConfiguration instead? Also this means anyone could add a different store impl this way, which seems a bit out of scope for the `connectSource` method currently.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/HotRodTargetMigrator.java [hunk 5]
```diff
@@ -112,4 +113,16 @@ public void disconnectSource(Cache<Object, Object> cache) throws CacheException
       });
       CompletableFuture.allOf(future).join();
    }
+
+   @Override
+   public void connectSource(Cache<Object, Object> cache, PersistenceConfiguration configuration) {
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/HotRodTargetMigrator.java:118]: I think this should return `CompletionStage<Void>` and do the join in the caller as necesary. Then in the rest server we don't need to use a blocking thread and can use this stage. The JMX method can just block.

**Review by wburns — CHANGES_REQUESTED** [review]
I still need to let the approach here stew a bit, but we need to for sure update all references to the stores of the persistence configuration needs to be updated. A lot of places assume we never change the number of stores at runtime so they will cache such things, not just the installation of interceptors.

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java [hunk 6]
```diff
@@ -323,11 +289,79 @@ private AsyncInterceptorChain buildInterceptorChain() {
...
+      if (configuration.persistence().usingStores()) {
+         addPersistenceInterceptors(interceptorChain, configuration, configuration.persistence());
+      }
+
       log.trace("Finished building default interceptor chain.");
       buildCustomInterceptors(interceptorChain, configuration.customInterceptors());
       return interceptorChain;
    }
 
+   public void addPersistenceInterceptors(AsyncInterceptorChain interceptorChain, Configuration cacheConfiguration, PersistenceConfiguration persistenceConfiguration) {
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java:302]: The issue is, when adding the persistence interceptor at runtime, it is hard to know the specific spot without calculating it, and the logic is inside ```buildInterceptorChain```, so I extracted the logic to be able to reuse it. 
I am open to suggestions to avoid this refactoring, is there a place in the interceptor stack where I can safely insert the persistence interceptors after the fact?

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:628]: Yes, as it stands it is too generic. Probably it should allow one 1 ```StoreConfiguration``` and at the API level (REST) it should validate it is a remote-store

[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:628]: > I am not a fan of reusing this method here. It has other purposes such as starting up the availability thread, which I think will be a problem here since by default that is enabled.

It was a actually the simplest refactor without changing the locking and blocking logic (```start``` was blocking before). But I get your point, it doing too much

File: core/src/main/java/org/infinispan/factories/EmptyConstructorNamedCacheFactory.java [hunk 7]
```diff
@@ -122,14 +121,11 @@ public Object construct(String componentName) {
...
-            }
-            return realPersistenceManager;
+         PersistenceManagerImpl persistenceManager = new PersistenceManagerImpl();
+         if (configuration.persistence().passivation()) {
+            return new PassivationPersistenceManager(persistenceManager);
          }
-         return new PersistenceManagerStub();
+         return persistenceManager;
       } else if (componentName.equals(PassivationManager.class.getName())) {
          return new PassivationManagerImpl();
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/factories/EmptyConstructorNamedCacheFactory.java:130]: Right. So currently it is possible to remove stores dynamically, does it mean the ```PassivationManager``` will keep in ```enabled``` state when it shouldn't?

File: documentation/src/main/asciidoc/topics/config_examples/remote-store.json (hunk 4 above)
Comment (gustavonalle) [reply documentation/src/main/asciidoc/topics/config_examples/remote-store.json:2]: As I wrote above, at REST API level, it probably makes sense to allow 1 and only 1 ```remote-store``` while at PM level it would allow only 1 ```StoreConfiguration```

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/HotRodTargetMigrator.java (hunk 5 above)
Comment (gustavonalle) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/HotRodTargetMigrator.java:118]: I didn't worry too much about this method since it a once in a lifetime of the server call, but sure, it'd be better to avoid blocking, I will change it

**Comment by gustavonalle** [comment]
Updated

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java [hunk 8]
```diff
@@ -166,6 +173,55 @@ public Invocations getInvocations() {
...
+
+      ConfigurationBuilder cfgBuilder = new ConfigurationBuilder();
+      PersistenceConfigurationBuilder persistenceConfigurationBuilder = cfgBuilder.persistence();
+      RemoteStoreConfigurationBuilder storeConfigurationBuilder = persistenceConfigurationBuilder
+            .addStore(RemoteStoreConfigurationBuilder.class);
+      invocationHelper.getJsonReader().readJson(storeConfigurationBuilder, storeConfig);
+
+      Configuration configuration = cfgBuilder.build();
+      PersistenceConfiguration persistence = configuration.persistence();
+      if (persistence.stores().size() != 1) {
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:209]: It doesn't look like this is checked to ensure there is only a remote store? Or is that done somewhere else?

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java [hunk 9]
```diff
@@ -51,6 +51,25 @@
...
+
+   /**
+    * Add a {@link StoreChangeListener} to be notified when a store is added or removed dynamically.
+    */
+   void addStoreListener(StoreChangeListener listener);
+
+   void removeStoreListener(StoreChangeListener listener);
+
+   interface StoreChangeListener {
+      void storeChanged(PersistenceStatus persistenceStatus);
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java:70]: It kinda needs additional synchronization depending on the implementation, it can be as simple as using volatiles, or extra locks to guarantee atomicity with multiple operations.

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java (hunk 8 above)
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:209]: This check is redundant, will get rid of it

**Comment by gustavonalle** [comment]
updated

**Comment by gustavonalle** [comment]
Rebased

**Comment by ryanemerson** [comment]
I resolved the Log conflict locally and pushed to main. Thanks @gustavonalle

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java]: I would have preferred to always add the persistence interceptors immediately after `EntryWrappingInterceptor`, even if it changed the order a bit. The advantage is that when a module adds another interceptor before EWI, the index is no longer valid, but you can always look up EWI and add the persistence interceptors after it.

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java (hunk 3 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]: Actually you don't need the write lock to modify the interceptors: a) components other than PMI could be modifying it, and b) `AsyncInterceptorChainImpl.addInterceptorAfter()` does its own locking, so you don't need any locking if you anchor everything on EntryWrappingInterceptor.

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java]: That looks good @danberindei I had it initially like this, but was unsure if it'd cause any issue not covered by the testsuite. I will send another PR

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java (hunk 3 above)
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]: I think I will leave the lock scope as it is. This method is far from being a hot spot, will be called likely never or 1 time per cache in the cluster lifetime

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java]: https://github.com/infinispan/infinispan/pull/9506
